polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -75,93 +76,264 @@ void Uniformity::print(raw_ostream &os) const { os << *this; }
+                                       AliasAnalysis &aliasAnalysis)
+    : SparseDataFlowAnalysis<UniformityLattice>(solver) {
+  // Load the reaching definition analysis (and the analysis it depends on).
+  // Reaching definition information are required by this analysis to reason
+  // about the uniformity of values loaded from memory.
+  internalSolver.load<DeadCodeAnalysis>();
+  internalSolver.load<SparseConstantPropagation>();
+  internalSolver.load<polygeist::ReachingDefinitionAnalysis>(aliasAnalysis);
+}
+
+LogicalResult UniformityAnalysis::initialize(Operation *top) {
+  // Run the dataflow analysis loaded in the internal solver.
+  if (failed(internalSolver.initializeAndRun(top)))
+    return failure();
+
+  return SparseDataFlowAnalysis::initialize(top);
+}
+
 void UniformityAnalysis::visitOperation(
     Operation *op, ArrayRef<const UniformityLattice *> operands,
     ArrayRef<UniformityLattice *> results) {
   LLVM_DEBUG(llvm::dbgs() << "UA: Visiting operation: " << *op << "\n");
 
-  // If the lattice on any operand is uninitialized, bail out.
+  // If the lattice on any operand isn't yet initialized, bail out.
   if (llvm::any_of(operands, [](const UniformityLattice *lattice) {
         return lattice->getValue().isUninitialized();
       })) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "Operand(s) uniformity not yet initialized\n");
     return;
   }
 
-  // These operations always yield non-uniform result(s).
-  if (isa<sycl::SYCLItemGetIDOp, sycl::SYCLNDItemGetGlobalIDOp>(op)) {
-    for (UniformityLattice *result : results)
-      propagateIfChanged(result, result->join(Uniformity::getNonUniform()));
-    return;
+  // Operations that always yield non-uniform result(s).
+  if (op->hasTrait<OpTrait::ResultsNonUniform>()) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "Operation yields non-uniform result(s)\n");
+    return propagateAllIfChanged(results, Uniformity::getNonUniform());
   }
 
-  auto anyOperandUniformityIs = [&](Uniformity::Kind kind) {
-    return llvm::any_of(operands, [&](const UniformityLattice *lattice) {
-      return lattice->getValue().getKind() == kind;
-    });
-  };
-
   // If any operand uniformity is unknown/non-uniform the result(s) are also
   // unknown/non-uniform.
-  if (anyOperandUniformityIs(Uniformity::Kind::Unknown)) {
-    setAllToEntryStates(results);
-    return;
+  if (anyOfUniformityIs(operands, Uniformity::Kind::Unknown)) {
+    LLVM_DEBUG(llvm::dbgs() << "Operand(s) uniformity is unknown\n");
+    return propagateAllIfChanged(results, Uniformity::getNonUniform());
   }
-  if (anyOperandUniformityIs(Uniformity::Kind::NonUniform)) {
-    for (UniformityLattice *result : results)
-      propagateIfChanged(result, result->join(Uniformity::getNonUniform()));
-    return;
+  if (anyOfUniformityIs(operands, Uniformity::Kind::NonUniform)) {
+    LLVM_DEBUG(llvm::dbgs() << "Operand(s) are non-uniform\n");
```

Comment:
just to be consistent with the previous debug line.
```suggestion
    LLVM_DEBUG(llvm::dbgs() << "Operand(s) uniformity is non-uniform\n");
```